prototype/mobile_endpoint/dao.py

```diff
@@ -0,0 +1,238 @@
+
+from mobile_endpoint.exceptions import IllegalCaseId
+from mobile_endpoint.form.models import XFormInstance
+from mobile_endpoint.models import db, Synclog, FormData, CaseData, CaseIndex
+from mobile_endpoint.utils import chunked, get_with_lock
+
+
+def to_generic(fn):
+    def _wrap(obj):
+        if hasattr(obj, 'to_generic'):
+            return obj.to_generic()
+        else:
+            return obj
+
+    @functools.wraps(fn)
+    def _inner(*args, **kwargs):
+        obj = fn(*args, **kwargs)
+        if isinstance(obj, (list, tuple)):
+            return [_wrap(ob) for ob in obj]
+        elif isinstance(obj, types.GeneratorType):
+            return (_wrap(ob) for ob in obj)
+        else:
+            return _wrap(obj)
+
+    return _inner
+
+class AbsctractDao(object):
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def commit(self, xform, cases):
+        """
+        Commit the transaction
+        """
+        pass
+
+    @abstractmethod
+    def get_synclog(self, id):
+        pass
+
+    @abstractmethod
+    def get_form(self, id):
+        pass
+    
+    @abstractmethod
+    def get_case(self, id, lock=False):
+        pass
+
+    @abstractmethod
+    def case_exists(self, id):
+        pass
+
+    @abstractmethod
+    def iter_cases(self, case_ids, chunksize=100, ordered=True):
+        pass
+
+    @abstractmethod
+    def get_reverse_indexed_cases(self, domain, case_ids):
+        pass
+
+
+class SQLDao(AbsctractDao):
+    def commit(self, xform, cases):
+        def to_sql(doc):
+            if isinstance(doc, XFormInstance):
+                return FormData.from_generic(doc)
+            elif isinstance(doc, CommCareCase):
+                return CaseData.from_generic(doc)
+
+        def get_indices():
+            for case in cases:
+                for index in case.indices:
+                    yield CaseIndex.from_generic(index, case.id)
+
+        new_form, xform_sql = to_sql(xform)
+        case_docs = map(to_sql, cases)
+
+        indices = list(get_indices())
+
+        for new_case, case_sql in case_docs:
+            case_sql.forms.append(xform_sql)
+
+        combined = [(new_form, xform_sql)] + case_docs + indices
+        for is_new, doc in combined:
+            if is_new:
+                db.session.add(doc)
+
+        db.session.commit()
+
+    @to_generic
+    def get_synclog(self, id):
+        return Synclog.query.get(id)
+
+    @to_generic
+    def get_form(self, id):
+        return FormData.query.get(id)
+
+    @to_generic
+    def get_case(self, id, lock=False):
+        if lock:
+            return get_with_lock('case_lock_{}'.format(id), lambda: CaseData.query.get(id))
+        else:
+            None, CaseData.query.get(id)
+
+    def case_exists(self, id):
+        return CaseData.query.filter_by(id=id).exists()
+
+    @to_generic
+    def iter_cases(self, case_ids, chunksize=100, ordered=False):
+        for doc_ids in chunked(case_ids, chunksize):
```

Comment:
I think for one thing, streaming from couchdb doesn't work awesome, because you need to json decode the entire payload to even look at the first element. I don't know if this is actually empirically the main problem, but it does mean that you have to wait for the entire payload and store it in memory to even get started.

I'm fairly certain that postgres and the related python libraries deal with this way better and let you cursor through data without loading all of it into memory first. In this case though, it looks like you're passing in all the ids to the query, and if there's a maximum query size (it that a thing?) that'd be an issue.